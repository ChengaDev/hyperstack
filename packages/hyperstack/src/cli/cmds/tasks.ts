/* eslint-disable @typescript-eslint/no-var-requires */
export default {
  command: 'tasks',
  describe: 'run a task',
  aliases: ['t'],
  builder: {
    list: {
      default: false,
      boolean: true,
      describe: 'show task list',
    },
  },
  async handler(argv: any) {
    const { root } = argv
    const { useTasks } = require('../../')
    return useTasks(
      {
        root,
      },
      { ...argv, _: argv._.splice(1) } // remove 'tasks' command
    )
  },
}
