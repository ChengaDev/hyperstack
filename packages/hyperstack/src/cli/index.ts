/* eslint-disable @typescript-eslint/no-var-requires */
import yargs from 'yargs'

export const cli = async (root: string, argv: string[]) => {
  return await yargs(argv.slice(2))
    .config({ root })
    .command(require('./cmds/migrate').default)
    .command(require('./cmds/generate').default)
    .command(require('./cmds/seed').default)
    .command(require('./cmds/routes').default)
    .command(require('./cmds/start').default)
    .command(require('./cmds/portal').default)
    .command(require('./cmds/tasks').default)
    .demandCommand()
    .help().argv
}
