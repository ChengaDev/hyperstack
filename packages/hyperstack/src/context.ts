import path from 'path'
import createDebug from 'debug'
import L from 'lodash'
import type { Config, ContextStore } from '@hyperstackjs/typings'
import { consts, setters } from './keys'
import type { Context } from './types'
import { printJobCounts, printRoutesV2 } from './utils'

const debug = createDebug('hyperstack:context')

class InMemContext implements Context {
  private __store: ContextStore = {}
  private __config: Config | null = null
  clear = () => {
    this.__store = {}
    this.__config = null
  }

  baseUrl = () => this.__store.controllers!.baseUrl
  logger = () => this.__store.logger
  models = () => this.__store.models!.models
  mailers = () => this.__store.workers!.mailers
  workers = () => this.__store.workers!.workers
  controllers = () => this.__store.controllers!.controllers
  environment = () => this.__store.app!.environmentFile
  helpers = {
    printRoutes: printRoutesV2,
    printJobCounts,
  }

  update = (updater: string, d: Partial<ContextStore>) => {
    debug('[%o] context update: %o', updater, d)
    L.merge(this.__store, d)
  }

  setConfig = (c: Config) => {
    this.__config = c
  }

  store = () => this.__store
  config = () => {
    if (!this.__config) {
      throw new Error('configuration not loaded yet.')
    }
    return this.__config
  }

  initializerProps = <T>(key: string) =>
    L.get(this.__store.initializers?.props || {}, key) as T
}

const context = new InMemContext()

export { context }

export const isTest = () => process.env.NODE_ENV === 'test'
export const isProduction = () => process.env.NODE_ENV === 'production'
export const isDevelopment = () => !isTest() && !isProduction()

export const envMode = () =>
  isTest() ? 'test' : isProduction() ? 'production' : 'development'
export const getEnvironment = () =>
  L.replace(process.env.NODE_ENV || 'development', /[^A-Za-z0-9_-]/g, '')

export const getAppEnvironment = () =>
  process.env[consts.appEnvKey]
    ? L.replace(process.env[consts.appEnvKey]!, /[^A-Za-z0-9_-]/g, '')
    : getEnvironment()

export const resolveAppRoot = (app?: string) =>
  app || path.join(process.cwd(), '/src')

export const tieApp = (context: Context, app?: string) => {
  const appRoot = resolveAppRoot(app)
  context.update(setters.context, {
    app: {
      mode: envMode(),
      environmentFile: getAppEnvironment(),
      root: appRoot,
    },
  })
}
