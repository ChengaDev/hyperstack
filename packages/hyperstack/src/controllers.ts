import path from 'path'
import fs from 'fs'
import type { ServerOptions } from '@hyperstackjs/hypercontroller'
import { createServer } from '@hyperstackjs/hypercontroller'
import type { Context } from './types'
import { setters } from './keys'
import { requireModule } from './utils'

const VAN_HALEN_PORT = 5150
const resolvePort = () => process.env.PORT || VAN_HALEN_PORT
const createApp = ({
  controllers,
  initializers,
  logging,
  staticFolder,
  indexCatchAll,
  forceHttps,
  gzip,
  cookieSecret,
  sendValidationErrors,
}: { controllers: any; initializers: any } & Partial<ServerOptions>) => {
  return createServer({
    controllers,
    opts: {
      logging,
      staticFolder,
      indexCatchAll,
      forceHttps,
      gzip,
      cookieSecret,
      sendValidationErrors,
      beforeMiddleware: (preapp: any) => {
        if (initializers && initializers.beforeMiddleware) {
          initializers.beforeMiddleware.forEach((init: (app: any) => void) =>
            init(preapp)
          )
        }
      },
      afterMiddleware: (postapp: any) => {
        if (initializers && initializers.afterMiddleware) {
          initializers.afterMiddleware.forEach((init: (app: any) => void) =>
            init(postapp)
          )
        }
      },
      beforeControllers: (preapp: any) => {
        if (initializers && initializers.beforeControllers) {
          initializers.beforeControllers.forEach((init: (app: any) => void) =>
            init(preapp)
          )
        }
      },
      afterControllers: (postapp: any) => {
        if (initializers && initializers.afterControllers) {
          initializers.afterControllers.forEach((init: (app: any) => void) =>
            init(postapp)
          )
        }
      },
    },
  })
}

export const getControllerConfig = (context: Context) => {
  const store = context.store()
  const appRoot = store.app!.root
  const logger = store.logger
  const middleware = store.logging!.middleware
  const initializers = store.initializers
  const controllerPath = path.join(appRoot, 'app', 'controllers')

  const {
    serveStatic,
    indexCatchAll,
    forceHttps,
    gzip,
    cookieSecret,
    sendValidationErrors,
    json,
    helmet,
    urlencoded,
    baseUrl: cBaseUrl,
  } = context.config().controllers || {}

  const baseUrl = cBaseUrl || `http://localhost:${resolvePort()}`

  if (!fs.existsSync(controllerPath)) {
    throw new Error(`tie controllers: no app controller in ${controllerPath}`)
  }
  const { AppController } = requireModule(controllerPath)
  if (!AppController) {
    throw new Error(`No 'AppController' export was found in ${controllerPath}`)
  }

  const staticFolder = serveStatic ? path.join(appRoot, 'public') : undefined
  return {
    controllers: AppController.controllers,
    initializers,
    logging: { logger, middleware },
    staticFolder,
    indexCatchAll,
    forceHttps,
    gzip,
    cookieSecret,
    sendValidationErrors,
    helmet,
    json,
    urlencoded,
    baseUrl,
  }
}

export const tieControllers = async (context: Context) => {
  const appConfig = getControllerConfig(context)
  const app = createApp(appConfig)

  context.update(setters.controllers, {
    controllers: {
      numControllers: appConfig.controllers.length,
      baseUrl: appConfig.baseUrl,
      controllers: appConfig.controllers,
      port: resolvePort(),
      app,
      didTieControllers: true,
    },
  })
}
