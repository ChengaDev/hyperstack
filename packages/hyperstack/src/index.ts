// expose the _right_ surface-area for users

export * from './tie'
export { initializer } from './initializers'
export { seed } from './models'
export { task } from './tasks'
export { isDevelopment, isProduction, context } from './context'
export { cli } from './cli'
export type {
  Context,
  Initializer,
  MigrationsMode,
  WorkersMode,
  Task,
  App,
  Config,
} from './types'

export {
  Schema,
  HyperModel,
  HyperModelError,
  ModelAssertionError,
  hashPassword,
  verifyPassword,
  generateHexToken,
} from '@hyperstackjs/hypermodel'

export {
  ClassErrorMiddleware,
  ClassMiddleware,
  Controller,
  Delete,
  ErrorMiddleware,
  Get,
  HttpResponse,
  HttpResponseBadRequest,
  HttpResponseClientError,
  HttpResponseConflict,
  HttpResponseCreated,
  HttpResponseForbidden,
  HttpResponseInternalServerError,
  HttpResponseMethodNotAllowed,
  HttpResponseMovedPermanently,
  HttpResponseNoContent,
  HttpResponseNotFound,
  HttpResponseNotImplemented,
  HttpResponseOK,
  HttpResponseRedirect,
  HttpResponseRedirection,
  HttpResponseServerError,
  HttpResponseSuccess,
  HttpResponseTooManyRequests,
  HttpResponseUnauthorized,
  Middleware,
  ParseBody,
  Patch,
  Post,
  Put,
  ValidateBody,
  ValidationError,
  authWithJWT,
  err,
  notfound,
  ok,
  permits,
  requires,
  unauthorized,
  Request,
  Response,
  NextFunction,
} from '@hyperstackjs/hypercontroller'

export {
  HyperWorker,
  Mailer,
  queueAs,
  producerOnly,
} from '@hyperstackjs/hyperworker'
