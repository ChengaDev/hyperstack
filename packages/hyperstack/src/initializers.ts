import path from 'path'
import L from 'lodash'
import createDebug from 'debug'
import { load } from './utils'
import type {
  Context,
  InitializerCreator,
  LifecycleHooks,
  LifecycleMap,
} from './types'
import { setters } from './keys'
const debug = createDebug('hyperstack:initializers')

export const initializer = (fn: InitializerCreator) => {
  return fn
}

export const getInitializers = (appRoot: string): InitializerCreator[] => {
  const initpath = path.join(appRoot, 'config/initializers/*.[tj]s')
  return (load(initpath) as InitializerCreator[]) || []
}

export const tieInitializers = async (context: Context) => {
  const appRoot = context.store().app!.root
  const initializerCreators = getInitializers(appRoot)
  const initializers: LifecycleHooks[] = []
  for (const initializerCreator of initializerCreators) {
    initializers.push(await initializerCreator(context))
  }

  // into initializermap, array initializers per lifecycle hook:
  // [{beforeApp:}, {beforeApp:}] -> { beforeApp: [..], ... afterApp: [..]}
  const inits = L.mergeWith({}, ...initializers, (v1: any, v2: any) =>
    L.isUndefined(v1) ? [v2] : [...v1, v2]
  ) as LifecycleMap

  inits.beforeApp && inits.beforeApp.forEach((init) => init())

  const props = {}
  inits.provideProps &&
    inits.provideProps.forEach((provide) => {
      const p = provide()
      debug('providing props: %o', p)
      L.merge(props, p)
    })

  context.update(setters.initializers, {
    initializers: { initializers: inits, props },
  })
}
