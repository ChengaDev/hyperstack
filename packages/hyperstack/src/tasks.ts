import path from 'path'
import L from 'lodash'
import { bold, underline, white, yellow } from 'colorette'
import { loadMap } from './utils'
import type { Context, Task, TaskFn, TaskMap } from './types'
import { setters } from './keys'

export const task = (desc: string, fn: TaskFn): Task => {
  return {
    description: desc,
    exec: fn,
  }
}

export const taskmapTable = (tasksRoot: string, taskmap: TaskMap) => {
  return `\n${bold(white('Tasks'))}\n\n${L.map(
    taskmap,
    (v, k) => `${underline(yellow(k))}: ${v.description}`
  ).join('\n')}`
}
export const printTasks = (
  tasksRoot: string,
  taskmap: TaskMap,
  printfn = console.log // eslint-disable-line no-console
) => {
  printfn(taskmapTable(tasksRoot, taskmap))
}
export const loadTasks = (tasksPath: string) => {
  const tasks = path.join(tasksPath, '**/*.[tj]s')
  return loadMap(tasks) || {}
}
export const taskfileToName = (tasksRoot: string, k: string) =>
  path
    .normalize(k)
    .replace(tasksRoot, '')
    .replace(/^[\\/]/, '')
    .replace(/\.[tj]s$/, '')

export const createTasks = (tasksRoot: string): Record<string, Task> => {
  const m = loadTasks(tasksRoot)
  const taskmap = L.mapKeys(m, (_v, k) => taskfileToName(tasksRoot, k))
  return taskmap
}

export const getTasks = (context: Context) => {
  const store = context.store()
  return {
    root: store.tasks!.root,
    map: store.tasks!.map,
  }
}

export const getTaskmap = (context: Context) => {
  return getTasks(context).map
}

export const tieTasks = (context: Context) => {
  const appPath = context.store().app!.root
  const customTasksRoot = context.config().tasks?.root
  const tasksRoot = customTasksRoot
    ? path.join(appPath, customTasksRoot)
    : path.join(appPath, 'lib', 'tasks')
  const taskMap = createTasks(tasksRoot)
  context.update(setters.tasks, {
    tasks: {
      root: tasksRoot,
      map: taskMap,
    },
  })
}
