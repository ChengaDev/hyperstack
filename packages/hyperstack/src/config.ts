import path from 'path'
import L from 'lodash'
import createDebug from 'debug'
import type { Config } from '@hyperstackjs/typings'
import { getAppEnvironment } from './context'
import { requireModule } from './utils'
import type { ConfigPaths, Context } from './types'
const debug = createDebug('hyperstack:config')

export const resolveConfig = ({ exactPath, appPath }: ConfigPaths) => {
  if (exactPath) {
    return exactPath
  }
  if (!appPath) {
    throw new Error('app path is empty')
  }
  return path.join(appPath, `config/environments/${getAppEnvironment()}`)
}
export const loadConfig = async (paths: ConfigPaths) => {
  const resolved = resolveConfig(paths)

  const mod = requireModule(resolved)

  // configs are function providing a dict.
  // it's a function because merging modifies the provided dict, and we
  // don't want to fiddle with module cache in case we want live reloads or tests to work
  const config = (await mod()) as Config

  debug('config before overrides is %o', config)
  if (paths.configOverrides) {
    L.merge(config, paths.configOverrides)
  }
  return { config, resolvedConfigPath: resolved }
}

export const tieConfig = async (
  context: Context,
  { exactPath, configOverrides }: ConfigPaths = {}
) => {
  const appPath = context.store().app!.root
  const { config, resolvedConfigPath } = await loadConfig({
    configOverrides,
    exactPath,
    appPath,
  })
  debug('loaded from %o', resolvedConfigPath)
  debug('config is %o', config)
  debug('overrides %o', configOverrides)
  context.setConfig(config)
}
