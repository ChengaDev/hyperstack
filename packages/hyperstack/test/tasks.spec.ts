import path from 'path'
import L from 'lodash'
import { context } from '../src/context'
import { getTasks, taskfileToName, taskmapTable, tieTasks } from '../src/tasks'
describe('tasks', () => {
  it('converts to name', () => {
    ;[
      ['/usr/bin/foo/bar/lib', '/usr/bin/foo/bar/lib/my/task.ts', 'my/task'],
      ['/usr/bin/foo/bar/lib', '/usr/bin/foo/bar/lib/task.ts', 'task'],
      ['/app/lib', '/app/lib/task.js', 'task'],
      [
        '/usr/bin/foo/bar/lib',
        '/usr/bin/foo/bar/lib/my//foo/bar/task.ts',
        'my/foo/bar/task',
      ],
    ].forEach((p) => {
      expect(taskfileToName(p[0], p[1])).toEqual(p[2])
    })
  })
  it('ties tasks', () => {
    context.clear()
    const appRoot = path.join(__dirname, 'fixtures', 'app-e2e')
    context.update('test', {
      app: {
        root: appRoot,
        mode: '',
        environmentFile: '',
      },
    })

    context.setConfig({})
    tieTasks(context)
    const tasks = getTasks(context)
    tasks.root = tasks.root.replace(appRoot, '')
    expect(tasks).toMatchSnapshot()

    context.clear()
    context.update('test', {
      app: {
        root: appRoot,
        mode: '',
        environmentFile: '',
      },
    })
    context.setConfig({})

    L.set(context.config(), 'tasks.root', 'custom-tasks-root')
    tieTasks(context)
    const tasks2 = getTasks(context)
    tasks2.root = tasks2.root.replace(appRoot, '')
    expect(tasks2).toMatchSnapshot()
    expect(taskmapTable(tasks2.root, tasks.map)).toMatchSnapshot()
  })
})
