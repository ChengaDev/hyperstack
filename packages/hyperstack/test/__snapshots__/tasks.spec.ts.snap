// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tasks ties tasks 1`] = `
{
  "map": {
    "weekly-email": {
      "description": "send a weekly email.",
      "exec": [Function],
    },
  },
  "root": "/lib/tasks",
}
`;

exports[`tasks ties tasks 2`] = `
{
  "map": {
    "foo-email": {
      "description": "custom task",
      "exec": [Function],
    },
  },
  "root": "/custom-tasks-root",
}
`;

exports[`tasks ties tasks 3`] = `
"
[1m[37mTasks[39m[22m

[4m[33mweekly-email[39m[24m: send a weekly email."
`;
